design-proposals/object-graph.md

```diff
@@ -0,0 +1,245 @@
+## **Overview**
+
+This is a proposal to include an `Object Graph` API in KubeVirt to represent VM and VMI dependencies and their relationships.
+
+## **Motivation**
+
+As new features continue to be added to KubeVirt, the graph of objects related to VMs (DataVolumes, PersistentVolumeClaims, InstanceTypes, Preferences, Secrets, ConfigMaps, etc.) continues to expand. Identifying all the objects that a VM depends on for tasks like backup, disaster recovery, or migration can be error-prone. We should simplify this process for users and partners by creating an authoritative way to retrieve a structured object graph.
```

Comment:
> We should simplify this process

Would you add a few word about the current process?
Could it be that we already use `ownerReferences` to reflect this information? It would be a problem to duplicate the information.

Reply:
It's not like there’s a "standardized" process, currently backup or migration solutions have to manually determine and include the objects they believe are related to a VM, without access to an authoritative source of truth.

Regarding ownerReferences, there are several VM dependencies that may lack them, like PVCs or DVs referenced in the Volumes section but not created using datavolumeTemplates.

Reply:
Thanks. If it is not standard, it is even more important to review the current process. We need to see if we can phase it out when introducing a new API, as we do not want to have two possibly-contradicting processes.

Reply:
Hey @dankenigsberg, sorry for the late response. Yeah I agree, for example, we've been doing something similar in the kubevirt velero plugin and we plan to deprecate the current method in favor of this structured graph. @awels is also planning to integrate this in his cluster migration feature, so this would provide a unified standard instead of duplicating logic across implementations. We are not aware if other software is using its own thing to fetch the vm dependencies, but the idea is to provide this api as the single source of truth.